### PR TITLE
Add iterators module

### PR DESCRIPTION
**Pull request by marler8997**
Add 2 new public functions and an `iterators` module.

`argsIterator`, returns an iterator that loops through its arguments

`chain`, accepts sub-iterators and loops through them all as if they were a single iterator

This PR may need to change depending on the fate of https://github.com/ziglang/zig/pull/2782

### CODE REVIEW

File: std/meta.zig [hunk 1]
```diff
@@ -117,6 +117,44 @@ test "std.meta.Child" {
     testing.expect(Child(promise->u8) == u8);
 }
 
+/// Given an array/pointer type, return the slice type `[]Child`. Preserves `const`.
```
Comment (marler8997) [inline std/meta.zig]: These functions will be considered in another PR (https://github.com/ziglang/zig/pull/2782).  For now I just include a version of them to get the tests to work.

**Comment by andrewrk** [comment]
please re-open when you have something ready for review

**Comment by jayschwa** [comment]
Hi @marler8997, I was just about to take a crack at this same problem when I searched and found this PR. Is it capable of resurrection, or did it ultimately not jive with the language in its current state?

**Comment by marler8997** [comment]
I played around with the concept of iterators and landed on an implementation for ranges here: https://github.com/marler8997/zog/blob/master/range.zig

It was alot of experimentation to see what sort of concepts would work in Zig.  I think Zig is probably capable of having a high-level range/iterator library lke this.  I haven't really found many applications for it yet since I haven't had too many projects in Zig yet, and when I have had use for it, I didn't want to start depending on it yet without having a package manager.

But you are welcome to take anything I've done and learn or expand upon it, or just trying something completely different as well.
